contrib/cloud.google.com/go/pubsub.v1/pubsub_test.go

```diff
@@ -236,31 +254,75 @@ func TestPropagationNoPubsliherSpan(t *testing.T) {
 	}, spans[0].Tags())
 }
 
-func setup(t *testing.T) (context.Context, *pubsub.Topic, *pubsub.Subscription, mocktracer.Tracer, func()) {
-	assert := assert.New(t)
-	mt := mocktracer.Start()
+func TestNamingSchema(t *testing.T) {
+	genSpans := namingschematest.GenSpansFn(func(t *testing.T, serviceOverride string) []mocktracer.Span {
+		var opts []Option
+		if serviceOverride != "" {
+			opts = append(opts, WithServiceName(serviceOverride))
+		}
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
+		defer cancel()
+
+		topic, sub := setup(t)
+
+		_, err := Publish(ctx, topic, &pubsub.Message{Data: []byte("hello"), OrderingKey: "xxx"}, opts...).Get(ctx)
+		require.NoError(t, err)
 
+		err = sub.Receive(ctx, WrapReceiveHandler(sub, func(ctx context.Context, msg *pubsub.Message) {
+			msg.Ack()
+			cancel()
+		}, opts...))
+		require.NoError(t, err)
+
+		return mt.FinishedSpans()
+	})
+	assertOpV0 := func(t *testing.T, spans []mocktracer.Span) {
+		require.Len(t, spans, 2)
+		assert.Equal(t, "pubsub.publish", spans[0].OperationName())
+		assert.Equal(t, "pubsub.receive", spans[1].OperationName())
+	}
+	assertOpV1 := func(t *testing.T, spans []mocktracer.Span) {
+		require.Len(t, spans, 2)
+		assert.Equal(t, "gcp.pubsub.send", spans[0].OperationName())
+		assert.Equal(t, "gcp.pubsub.process", spans[1].OperationName())
+	}
+	serviceOverride := namingschematest.TestServiceOverride
+	wantServiceNameV0 := namingschematest.ServiceNameAssertions{
+		WithDefaults:             []string{"", ""},
+		WithDDService:            []string{"", ""},
+		WithDDServiceAndOverride: []string{serviceOverride, serviceOverride},
+	}
+	t.Run("ServiceName", namingschematest.NewServiceNameTest(genSpans, "", wantServiceNameV0))
+	t.Run("SpanName", namingschematest.NewOpNameTest(genSpans, assertOpV0, assertOpV1))
+}
+
+func setup(t *testing.T) (*pubsub.Topic, *pubsub.Subscription) {
```

Comment:
+1 on the `require`s here, and the `t.Cleanup`, but I'm confused about the removal of the context and mocktracer.

If you want a different timeout or something, we can either change it or make it a parameter.

Reply:
The motivation behind not returning the context is to not carry a context that already has a timeout and also to have access outside it to the `cancel` function, which can be used to speed up tests a little bit (the `sub.Receive` function returns when the context is canceled, so we can cancel when the work is finished instead of waiting for the timeout).

The mocktracer is just I didn't see much benefit of having it here compared to each test (just saving 2 lines) and I found a little bit confusing returning that many values from a single function. This is a little bit subjective, so if you still prefer the previous version let me know and I'll change 😄

Reply:
I prefer the previous version for 2 reasons:

1. It's better to have a context that already has a timeout. You can always modify the context with an additional, shorter timeout for tests that can benefit from it. Having `setup` take care of that means that test contexts are always going to timeout. Likewise, if you want a cancel function, you can add one yourself where you need it. Also, everywhere where you are creating a context, you're using the same timeout anyway. If you want a shorter timeout, you can just change it in the setup function. 

2. No reason to repeat the same code for creating a mock tracer over and over. Four is a lot of return values. I'm ok with it for test harnesses like this, but I'd also be ok returning a struct containing these instead if you really want to refactor.

Reply:
changed!